namesys/pubsub.go

```diff
@@ -0,0 +1,427 @@
+	return p.ps.Publish(topic, data)
+}
+
+// Resolve resolves a name through pubsub and default depth limit
+func (r *PubsubResolver) Resolve(ctx context.Context, name string) (path.Path, error) {
+	return r.ResolveN(ctx, name, DefaultDepthLimit)
+}
+
+// ResolveN resolves a name through pubsub with the specified depth limit
+func (r *PubsubResolver) ResolveN(ctx context.Context, name string, depth int) (path.Path, error) {
+	return resolve(ctx, r, name, depth, "/ipns/")
+}
+
+func (r *PubsubResolver) resolveOnce(ctx context.Context, name string) (path.Path, error) {
+	log.Debugf("PubsubResolve: resolve '%s'", name)
+
+	// retrieve the public key once (for verifying messages)
+	xname := strings.TrimPrefix(name, "/ipns/")
+	hash, err := mh.FromB58String(xname)
+	if err != nil {
+		log.Warningf("PubsubResolve: bad input hash: [%s]", xname)
+		return "", err
+	}
+
+	id := peer.ID(hash)
+	if r.host.Peerstore().PrivKey(id) != nil {
+		return "", errors.New("Cannot resolve own name through pubsub")
+	}
+
+	pubk := id.ExtractPublicKey()
+	if pubk == nil {
+		pubk, err = r.pkf.GetPublicKey(ctx, id)
+		if err != nil {
+			log.Warningf("PubsubResolve: error fetching public key: %s [%s]", err.Error(), xname)
+			return "", err
+		}
+	}
+
+	// the topic is /ipns/Qmhash
+	if !strings.HasPrefix(name, "/ipns/") {
+		name = "/ipns/" + name
+	}
+
+	r.mx.Lock()
+	// see if we already have a pubsub subscription; if not, subscribe
+	sub, ok := r.subs[name]
+	if !ok {
+		sub, err = r.ps.Subscribe(name)
+		if err != nil {
+			r.mx.Unlock()
+			return "", err
+		}
+
+		log.Debugf("PubsubResolve: subscribed to %s", name)
+
+		r.subs[name] = sub
+
+		ctx, cancel := context.WithCancel(r.ctx)
+		go r.handleSubscription(sub, name, pubk, cancel)
+		go bootstrapPubsub(ctx, r.cr, r.host, name)
+	}
+	r.mx.Unlock()
+
+	// resolve to what we may already have in the datastore
+	dsval, err := r.ds.Get(dshelp.NewKeyFromBinary([]byte(name)))
+	if err != nil {
+		if err == ds.ErrNotFound {
+			return "", ErrResolveFailed
+		}
+		return "", err
+	}
+
+	data := dsval.([]byte)
+	entry := new(pb.IpnsEntry)
+
+	err = proto.Unmarshal(data, entry)
+	if err != nil {
+		return "", err
+	}
+
+	// check EOL; if the entry has expired, delete from datastore and return ds.ErrNotFound
+	eol, ok := checkEOL(entry)
+	if ok && eol.Before(time.Now()) {
+		err = r.ds.Delete(dshelp.NewKeyFromBinary([]byte(name)))
+		if err != nil {
+			log.Warningf("PubsubResolve: error deleting stale value for %s: %s", name, err.Error())
+		}
+
+		return "", ErrResolveFailed
+	}
+
+	value, err := path.ParsePath(string(entry.GetValue()))
+	return value, err
+}
+
+// GetSubscriptions retrieves a list of active topic subscriptions
+func (r *PubsubResolver) GetSubscriptions() []string {
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	var res []string
+	for sub := range r.subs {
+		res = append(res, sub)
+	}
+
+	return res
+}
+
+// Cancel cancels a topic subscription
+func (r *PubsubResolver) Cancel(name string) {
```

Comment:
maybe this should return a bool for whether or not the named thing was actually removed? I'm concerned about the use case of `Cancel("thing that doesnt exist")`

Reply:
Shouldn't that be a noop?

Reply:
updated to return a boolean indicator in https://github.com/ipfs/go-ipfs/pull/4047/commits/2e6af50a12d445b673fec0121e1857dc7713b1c3